In [448]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import re
import warnings

warnings.filterwarnings("ignore")

In [450]:
df = pd.read_csv('usedCarListing.csv')
del df['Unnamed: 0']
print(df.shape)
df.head()

(9841, 19)


year     make    model                      sub_model           city state  \
0  2017     Ford   Fusion                         SE FWD     Blue Mound    TX   
1  2007      BMW        6               650i Convertible        Stanton    CA   
2  2019      Kia   Optima                             LX       Thornton    CO   
3  2017     Ford   Fusion                         SE FWD         Denver    NC   
4  2017  Hyundai  Elantra  SE 2.0L Sedan Automatic (alt)  Wesley Chapel    FL   

   mileage    price                    exterior_color      interior_color  \
0   32,253  $14,400  White Platinum Metallic Tri-Coat  Medium Light Stone   
1   59,664  $11,499           Mineral Silver Metallic                 Tan   
2    4,816  $18,586                       Ebony Black               Black   
3   99,515   $8,999                      Shadow Black               Black   
4  117,200   $6,109                Quartz White Pearl               Beige   

  mpg_city mpg_hwy                          engine transmission drive_type  \
0       21      32  1.5L Inline-4 Gas Turbocharged    Automatic        FWD   
1       16      22                    4.8L V-8 Gas    Automatic        RWD   
2       24      33               2.4L Inline-4 Gas    Automatic        FWD   
3       21      32               2.5L Inline-4 Gas    Automatic        FWD   
4       29      38               2.0L Inline-4 Gas    Automatic        FWD   

  fuel_type                                    popular_feature  \
0       Gas  ['Navigation', 'Backup Camera', 'Cruise Contro...   
1       Gas                                                 []   
2       Gas  ['Cruise Control', 'Lane Keep Assist', 'Blueto...   
3       Gas   ['Cruise Control', 'Bluetooth', 'Backup Camera']   
4       Gas                    ['Cruise Control', 'Bluetooth']   

                                     vehicle_history    cpo  
0  ['0', ' reported accidents', 'Fleet or Mixed U...  False  
1  ['0', ' reported accidents', 'Fleet or Mixed U...  False  
2  ['1', ' reported accidents', 'Personal or Rent...   True  
3  ['1', ' reported accidents', 'Fleet or Mixed U...  False  
4  ['1', ' reported accidents', 'Personal or Rent...  False

## create new features

In [451]:
# split vehicle history into four variables(num_accident, usage_type, title, num_owner)
split_data = df["vehicle_history"].str.split(",")
data = split_data.to_list()
names = ["1", "2", "3", "4", '5', '6']
new_df = pd.DataFrame(data, columns=names)
df['num_accident'] = new_df['1'].apply(lambda x:x[2:-1])
df['usage_type'] = new_df['3'].apply(lambda x:x[2:-1]) 
df['title'] = new_df['4'].apply(lambda x:x[2:-1])
df['num_owner'] = new_df['5'].apply(lambda x:x[2:-2])

# delete the history column
del df['vehicle_history']

df.head()

year     make    model                      sub_model           city state  \
0  2017     Ford   Fusion                         SE FWD     Blue Mound    TX   
1  2007      BMW        6               650i Convertible        Stanton    CA   
2  2019      Kia   Optima                             LX       Thornton    CO   
3  2017     Ford   Fusion                         SE FWD         Denver    NC   
4  2017  Hyundai  Elantra  SE 2.0L Sedan Automatic (alt)  Wesley Chapel    FL   

   mileage    price                    exterior_color      interior_color  \
0   32,253  $14,400  White Platinum Metallic Tri-Coat  Medium Light Stone   
1   59,664  $11,499           Mineral Silver Metallic                 Tan   
2    4,816  $18,586                       Ebony Black               Black   
3   99,515   $8,999                      Shadow Black               Black   
4  117,200   $6,109                Quartz White Pearl               Beige   

   ...                          engine transmission drive_type fuel_type  \
0  ...  1.5L Inline-4 Gas Turbocharged    Automatic        FWD       Gas   
1  ...                    4.8L V-8 Gas    Automatic        RWD       Gas   
2  ...               2.4L Inline-4 Gas    Automatic        FWD       Gas   
3  ...               2.5L Inline-4 Gas    Automatic        FWD       Gas   
4  ...               2.0L Inline-4 Gas    Automatic        FWD       Gas   

                                     popular_feature    cpo num_accident  \
0  ['Navigation', 'Backup Camera', 'Cruise Contro...  False            0   
1                                                 []  False            0   
2  ['Cruise Control', 'Lane Keep Assist', 'Blueto...   True            1   
3   ['Cruise Control', 'Bluetooth', 'Backup Camera']  False            1   
4                    ['Cruise Control', 'Bluetooth']  False            1   

               usage_type  title num_owner  
0      Fleet or Mixed Use  Clean         1  
1      Fleet or Mixed Use  Clean         3  
2  Personal or Rental Use  Clean         2  
3      Fleet or Mixed Use  Clean         2  
4  Personal or Rental Use  Clean         1  

[5 rows x 22 columns]

In [452]:
df.num_owner.unique(),df.num_accident.unique(),df.title.unique(),df.usage_type.unique()

(array(['1', '3', '2', '5', '4', '0', '6', '7', '8', 'alvage', '9', '13',
        'rame Damage'], dtype=object),
 array(['0', '1', '2', '3', '4', '5'], dtype=object),
 array(['Clean', 'Theft Recovered', 'Salvaged', 'Frame Damaged',
        'Theft Recovere', 'Lemon', 'Frame Damage'], dtype=object),
 array(['Fleet or Mixed Use', 'Personal or Rental Use'], dtype=object))

In [453]:
### Deal with title (combine theft recovered with theft recovere, combine frame damaged with frame damage)
df["title"]= df["title"].replace('Theft Recovere', "Theft Recovered") 
df["title"]= df["title"].replace('Frame Damage', "Frame Damaged") 
print(df.title.unique())

### Deal with num_owner that are strings
df["num_owner"]= df["num_owner"].replace('alvage', np.nan) 
df["num_owner"]= df["num_owner"].replace('rame Damage', np.nan) 
print(df.num_owner.unique())

['Clean' 'Theft Recovered' 'Salvaged' 'Frame Damaged' 'Lemon']
['1' '3' '2' '5' '4' '0' '6' '7' '8' nan '9' '13']


## check for missing values

In [454]:
df.isna().sum()

year               0
make               0
model              0
sub_model          0
city               0
state              0
mileage            0
price              0
exterior_color     0
interior_color     1
mpg_city           0
mpg_hwy            0
engine             0
transmission       0
drive_type         0
fuel_type          0
popular_feature    0
cpo                0
num_accident       0
usage_type         0
title              0
num_owner          5
dtype: int64

In [455]:
# categorical missing value, take the mode to fillna
df['interior_color'].fillna(df['interior_color'].value_counts().index[0], inplace=True)
# num_owern, numeric missing value, take the median to fillna
df['num_owner'].fillna(df['num_owner'].median(), inplace=True)
df.isna().sum()
# 

year               0
make               0
model              0
sub_model          0
city               0
state              0
mileage            0
price              0
exterior_color     0
interior_color     0
mpg_city           0
mpg_hwy            0
engine             0
transmission       0
drive_type         0
fuel_type          0
popular_feature    0
cpo                0
num_accident       0
usage_type         0
title              0
num_owner          0
dtype: int64

In [456]:
## look at mpg
print(df.mpg_city.unique())
print(df.mpg_hwy.unique())
## convert 'N' and 'cty' to nan.
df["mpg_city"]= df["mpg_city"].replace('N', np.nan) 
df["mpg_hwy"]= df["mpg_hwy"].replace('cty', np.nan) 
print(df.mpg_city.unique())
print(df.mpg_hwy.unique())

['21' '16' '24' '29' '31' '18' '17' '25' '23' '26' '19' '22' '15' 'N' '20'
 '14' '28' '27' '110' '30' '104' '13' '12' '101' '43' '44' '106' '34' '10'
 '95' '126' '85' '121' '32' '11' '9' '48' '122' '120' '41' '129' '40' '35'
 '38' '51' '49' '36' '66' '42' '33' '60' '124' '53' '128' '39' '91' '54'
 '47' '37' '8' '102']
['32' '22' '33' '38' '40' '25' '23' '30' '36' '31' '41' '29' '28' '34'
 '19' 'cty' '26' '21' '24' '37' '39' '99' '91' '18' '17' '93' '35' '44'
 '16' '92' '27' '15' '20' '81' '82' '103' '42' '45' '90' '108' '105' '102'
 '43' '48' '47' '66' '49' '51' '46' '109' '101' '100' '14' '50' '13' '110'
 '94' '12']
['21' '16' '24' '29' '31' '18' '17' '25' '23' '26' '19' '22' '15' nan '20'
 '14' '28' '27' '110' '30' '104' '13' '12' '101' '43' '44' '106' '34' '10'
 '95' '126' '85' '121' '32' '11' '9' '48' '122' '120' '41' '129' '40' '35'
 '38' '51' '49' '36' '66' '42' '33' '60' '124' '53' '128' '39' '91' '54'
 '47' '37' '8' '102']
['32' '22' '33' '38' '40' '25' '23' '30' '36' '31' '41'

In [457]:
# deal with mpg_city, mpg_hwy
# first convert mpg from string to numeric values
df['mpg_city'] = df['mpg_city'].astype(float)
df['mpg_hwy'] = df['mpg_hwy'].astype(float)

In [458]:
# since mpg is strongly correlated with vehicle fuel_type, especially for alternative fuel vehicles.
# Use vehicle with same fuel_type to fillna.
df[df['mpg_city'].isna()]['fuel_type'].unique()

array(['Gas', 'Hydrogen', 'Diesel', 'Plug-In', 'Hybrid'], dtype=object)

In [459]:
df[(df['fuel_type']=='Gas') & (df['mpg_city'].isna())].head()

year    make  model                      sub_model              city  \
40   2017    Ford  Super         XL Crew Cab 8' Bed 4WD         Lexington   
159  2003  HUMMER     H2                            SUV            Belton   
240  2004  HUMMER     H2                            SUV            McCook   
269  2017    Ford  Super  Lariat Crew Cab 6.75' Bed 4WD             Boise   
278  2004   Dodge    Ram   SLT Quad Cab Regular Bed 4WD  Colorado Springs   

    state  mileage    price exterior_color     interior_color  ...  \
40     NC   94,903  $29,987          White  Medium Earth Gray  ...   
159    TX  139,739  $11,900          Black              Wheat  ...   
240    IL  149,795   $9,950          White              Wheat  ...   
269    ID   70,634  $35,995       Race Red            Unknown  ...   
278    CO  159,551  $10,300           Gray              Black  ...   

           engine  transmission drive_type fuel_type popular_feature    cpo  \
40   6.2L V-8 Gas     Automatic        4WD       Gas              []  False   
159  6.0L V-8 Gas     Automatic        4WD       Gas              []  False   
240  6.0L V-8 Gas     Automatic        4WD       Gas              []  False   
269  6.2L V-8 Gas     Automatic        4WD       Gas              []  False   
278  5.7L V-8 Gas     Automatic        4WD       Gas              []  False   

    num_accident              usage_type  title num_owner  
40             0  Personal or Rental Use  Clean         1  
159            0      Fleet or Mixed Use  Clean         6  
240            2  Personal or Rental Use  Clean         1  
269            0  Personal or Rental Use  Clean         1  
278            0      Fleet or Mixed Use  Clean         3  

[5 rows x 22 columns]

In [460]:
# use fuel_type (electric, plug-in, hybrid, hydrogen) to fillna first. 
# deal with mpg_city first.
for i in range(len(df)):
    if (df['fuel_type'][i] == 'Hydrogen' and np.isnan(df['mpg_city'][i])):
        df['mpg_city'][i]= df[df['fuel_type']=='Hydrogen']['mpg_city'].mean()
    elif (df['fuel_type'][i] == 'Plug-In' and np.isnan(df['mpg_city'][i])):
        df['mpg_city'][i]=df[df['fuel_type']=='Plug-In']['mpg_city'].mean()
    elif (df['fuel_type'][i] == 'Hybrid' and np.isnan(df['mpg_city'][i])):
        df['mpg_city'][i]=df[df['fuel_type']=='Hybrid']['mpg_city'].mean()
    # The missing values for Diesel and Gas vehicles are more like gas-gazzlers. So I can't use mean or median
    # to replace the missing values. 
    # I decided to use the 95% quantile values to fillna
    elif (df['fuel_type'][i] == 'Gas' and np.isnan(df['mpg_city'][i])):
        df['mpg_city'][i]=df[df['fuel_type']=='Gas']['mpg_city'].quantile(q=0.95)
    elif (df['fuel_type'][i] == 'Diesel' and np.isnan(df['mpg_city'][i])):
        df['mpg_city'][i]=df[df['fuel_type']=='Diesel']['mpg_city'].quantile(q=0.95)

# then deal with mpg_hwy.      
for i in range(len(df)):
    if (df['fuel_type'][i] == 'Hydrogen' and np.isnan(df['mpg_hwy'][i])):
        df['mpg_hwy'][i]=df[df['fuel_type']=='Hydrogen']['mpg_hwy'].mean()
    elif (df['fuel_type'][i] == 'Plug-In' and np.isnan(df['mpg_hwy'][i])):
        df['mpg_hwy'][i]=df[df['fuel_type']=='Plug-In']['mpg_hwy'].mean()
    elif (df['fuel_type'][i] == 'Hybrid' and np.isnan(df['mpg_hwy'][i])):
        df['mpg_hwy'][i]=df[df['fuel_type']=='Hybrid']['mpg_hwy'].mean()        
    
    # For diesel vehicles with missing values, I found all diesel vehicles are missing the mpg_hwy. 
    elif (df['fuel_type'][i] == 'Gas' and np.isnan(df['mpg_hwy'][i])):
        df['mpg_hwy'][i]=df[df['fuel_type']=='Gas']['mpg_hwy'].mean()
    elif (df['fuel_type'][i] == 'Diesel' and np.isnan(df['mpg_hwy'][i])):
        df['mpg_hwy'][i]=df[df['fuel_type']=='Diesel']['mpg_hwy'].mean()        
        
df.head()

year     make    model                      sub_model           city state  \
0  2017     Ford   Fusion                         SE FWD     Blue Mound    TX   
1  2007      BMW        6               650i Convertible        Stanton    CA   
2  2019      Kia   Optima                             LX       Thornton    CO   
3  2017     Ford   Fusion                         SE FWD         Denver    NC   
4  2017  Hyundai  Elantra  SE 2.0L Sedan Automatic (alt)  Wesley Chapel    FL   

   mileage    price                    exterior_color      interior_color  \
0   32,253  $14,400  White Platinum Metallic Tri-Coat  Medium Light Stone   
1   59,664  $11,499           Mineral Silver Metallic                 Tan   
2    4,816  $18,586                       Ebony Black               Black   
3   99,515   $8,999                      Shadow Black               Black   
4  117,200   $6,109                Quartz White Pearl               Beige   

   ...                          engine  transmission drive_type fuel_type  \
0  ...  1.5L Inline-4 Gas Turbocharged     Automatic        FWD       Gas   
1  ...                    4.8L V-8 Gas     Automatic        RWD       Gas   
2  ...               2.4L Inline-4 Gas     Automatic        FWD       Gas   
3  ...               2.5L Inline-4 Gas     Automatic        FWD       Gas   
4  ...               2.0L Inline-4 Gas     Automatic        FWD       Gas   

                                     popular_feature    cpo num_accident  \
0  ['Navigation', 'Backup Camera', 'Cruise Contro...  False            0   
1                                                 []  False            0   
2  ['Cruise Control', 'Lane Keep Assist', 'Blueto...   True            1   
3   ['Cruise Control', 'Bluetooth', 'Backup Camera']  False            1   
4                    ['Cruise Control', 'Bluetooth']  False            1   

               usage_type  title num_owner  
0      Fleet or Mixed Use  Clean         1  
1      Fleet or Mixed Use  Clean         3  
2  Personal or Rental Use  Clean         2  
3      Fleet or Mixed Use  Clean         2  
4  Personal or Rental Use  Clean         1  

[5 rows x 22 columns]

In [461]:
df.isna().sum()

year               0
make               0
model              0
sub_model          0
city               0
state              0
mileage            0
price              0
exterior_color     0
interior_color     0
mpg_city           0
mpg_hwy            0
engine             0
transmission       0
drive_type         0
fuel_type          0
popular_feature    0
cpo                0
num_accident       0
usage_type         0
title              0
num_owner          0
dtype: int64

In [462]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9841 entries, 0 to 9840
Data columns (total 22 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   year             9841 non-null   int64  
 1   make             9841 non-null   object 
 2   model            9841 non-null   object 
 3   sub_model        9841 non-null   object 
 4   city             9841 non-null   object 
 5   state            9841 non-null   object 
 6   mileage          9841 non-null   object 
 7   price            9841 non-null   object 
 8   exterior_color   9841 non-null   object 
 9   interior_color   9841 non-null   object 
 10  mpg_city         9841 non-null   float64
 11  mpg_hwy          9841 non-null   float64
 12  engine           9841 non-null   object 
 13  transmission     9841 non-null   object 
 14  drive_type       9841 non-null   object 
 15  fuel_type        9841 non-null   object 
 16  popular_feature  9841 non-null   object 
 17  cpo           

## Convert some categorical variables into numeric
Numeric variable: year, mileage, price, mpg_city, mpg_hwy, num_accident, num_owner

In [463]:
# deal with mileage for the comma.
df['mileage'] = df['mileage'].apply(lambda x: ''.join(x.split(',')))
df['mileage'] = df['mileage'].astype(int)
# deal with price
df['price'] = df['price'].apply(lambda x: ''.join(x[1:].split(',')))
df['price'] = df['price'].astype(int)
# deal with num_accident and num_owner
df['num_accident'] = df['num_accident'].astype(int)
df['num_owner'] = df['num_owner'].astype(int)

In [464]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9841 entries, 0 to 9840
Data columns (total 22 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   year             9841 non-null   int64  
 1   make             9841 non-null   object 
 2   model            9841 non-null   object 
 3   sub_model        9841 non-null   object 
 4   city             9841 non-null   object 
 5   state            9841 non-null   object 
 6   mileage          9841 non-null   int64  
 7   price            9841 non-null   int64  
 8   exterior_color   9841 non-null   object 
 9   interior_color   9841 non-null   object 
 10  mpg_city         9841 non-null   float64
 11  mpg_hwy          9841 non-null   float64
 12  engine           9841 non-null   object 
 13  transmission     9841 non-null   object 
 14  drive_type       9841 non-null   object 
 15  fuel_type        9841 non-null   object 
 16  popular_feature  9841 non-null   object 
 17  cpo           

## drop duplicate rows

In [465]:
# There are totally 139 duplcated rows
print(df.duplicated().sum())
# drop duplicated rows and keep the first row.
print(df.shape)
df = df.drop_duplicates(keep='first')
print(df.shape)

139
(9841, 22)
(9702, 22)


## Deal with potential wrong data (mileage, price)

In [471]:
# Check very low mileage and very high mileage
display(df.sort_values(by='mileage',ascending=False).head())
df.sort_values(by='mileage',ascending=True).head()
## The data seems making sense to me. I won't drop any data from mileage here.

year        make    model                     sub_model  \
1346  2012         Ram     3500       SLT Crew Cab 8' Box 2WD   
9194  2012  Volkswagen    Jetta            TDI SportWagen DSG   
1916  2012        Ford    Super  Lariat Crew Cab 172" DRW 4WD   
8470  2004      Toyota  Sequoia                   Limited 4WD   
2427  2015   Chevrolet   Impala                  LTZ with 2LZ   

                 city state  mileage  price       exterior_color  \
1346        Charlotte    NC   484773  14995         Bright White   
9194        Kingsport    TN   445612   4900               Silver   
1916          Phoenix    AZ   366633  15888                White   
8470  East Landsdowne    PA   341000   3475               Pewter   
2427      San Antonio    TX   325926   8250  Silver Ice Metallic   

                            interior_color  ...  \
1346  Dark Slate/Medium Graystone Interior  ...   
9194                           Titan Black  ...   
1916                               Unknown  ...   
8470                                  Gray  ...   
2427               Jet Black/Dark Titanium  ...   

                                 engine  transmission drive_type fuel_type  \
1346  6.7L Inline-6 Diesel Turbocharged     Automatic        RWD    Diesel   
9194  2.0L Inline-4 Diesel Turbocharged     Automatic        FWD    Diesel   
1916       6.7L V-8 Diesel Turbocharged     Automatic        4WD    Diesel   
8470                       4.7L V-8 Gas     Automatic        4WD       Gas   
2427                       3.6L V-6 Gas     Automatic        FWD       Gas   

                                        popular_feature    cpo num_accident  \
1346          ['Remote Engine Start', 'Cruise Control']  False            1   
9194                                                 []  False            1   
1916                                                 []  False            0   
8470                                                 []  False            0   
2427  ['Moonroof', 'Front Cooled Seats', 'Cruise Con...  False            0   

                  usage_type  title num_owner  
1346      Fleet or Mixed Use  Clean         2  
9194  Personal or Rental Use  Clean         1  
1916      Fleet or Mixed Use  Clean         1  
8470  Personal or Rental Use  Clean         5  
2427  Personal or Rental Use  Clean         1  

[5 rows x 22 columns]

year           make     model                      sub_model  \
7003  2020  Mercedes-Benz       GLE                 GLE 350 4MATIC   
2729  2019  Mercedes-Benz  Sprinter                      4500 170"   
8554  2018          Dodge   Journey                         SE FWD   
880   2019  Mercedes-Benz  Sprinter                    3500XD 170"   
4521  2017           Ford     F-150  Lariat SuperCrew 5.5' Box 4WD   

             city state  mileage  price                    exterior_color  \
7003  Little Rock    AR        7  52878                       Polar White   
2729       Monroe    NC        8  28998                             White   
8554       Dallas    GA        8  14767              Redline 2 Coat Pearl   
880       Houston    TX       10  33995                             White   
4521   Somerville    NJ       10  31500  White Platinum Metallic Tri-Coat   

     interior_color  ...                          engine  transmission  \
7003          Black  ...  2.0L Inline-4 Gas Turbocharged     Automatic   
2729        Unknown  ...    3.0L V-6 Diesel Turbocharged     Automatic   
8554          Black  ...               2.4L Inline-4 Gas     Automatic   
880            Gray  ...    3.0L V-6 Diesel Turbocharged     Automatic   
4521          Black  ...       5.0L V-8 Gas Turbocharged     Automatic   

     drive_type fuel_type                                    popular_feature  \
7003        AWD       Gas                                                 []   
2729        RWD    Diesel                                                 []   
8554        FWD       Gas   ['Cruise Control', 'Bluetooth', 'Backup Camera']   
880         RWD    Diesel                                                 []   
4521        4WD       Gas  ['Moonroof', 'Front Heated Seats', 'Remote Eng...   

        cpo num_accident              usage_type  title num_owner  
7003  False            0  Personal or Rental Use  Clean         0  
2729  False            0  Personal or Rental Use  Clean         0  
8554  False            0  Personal or Rental Use  Clean         1  
880   False            0  Personal or Rental Use  Clean         1  
4521  False            0  Personal or Rental Use  Clean         1  

[5 rows x 22 columns]

In [476]:
# Check very low price and very high price
display(df.sort_values(by='price',ascending=False).head(10))
df.sort_values(by='price',ascending=True).head(10)
## The high and low prices seems making sense to me. Luxury used cars have very high price. 
## Old used cars have very low prices
## I won't drop any data from mileage here.

year         make       model      sub_model              city state  \
5513  2018      Bentley    Bentayga  W12 Signature     Downers Grove    IL   
6761  2018      Bentley    Bentayga       Mulliner  Westlake Village    CA   
6392  2005  Lamborghini  Murcielago          Coupe            Boerne    TX   
1898  2015  Rolls-Royce      Wraith            RWD       San Antonio    TX   
2647  2010      Ferrari         599          Coupe           Ontario    CA   
434   2008      Ferrari         599          Coupe            Boerne    TX   
436   2015      Ferrari  California    Convertible         Melbourne    FL   
6255  2008      Ferrari         599          Coupe          Marietta    GA   
1918  2012      Ferrari          FF      Hatchback        Pittsburgh    PA   
442   2008      Bentley       Azure    Convertible            Naples    FL   

      mileage   price         exterior_color     interior_color  ...  \
5513     4118  169992               Magnolia             Beluga  ...   
6761     9160  156491       Granite Metallic              Brown  ...   
6392     4565  148888          Grey Metallic                Red  ...   
1898    14077  141968         Infinity Black              Black  ...   
2647    30061  139500  Nero Daytona Metallic               Nero  ...   
434     18075  129888              Corsa Red     Interior Color  ...   
436      5835  128998  Nero Daytona Metallic             Sabbia  ...   
6255    16148  124900       Silverstone Gray     Interior Color  ...   
1918    25400  118075                   Gray                Tan  ...   
442     10393  109995           Beluga Black  Standard Interior  ...   

                          engine  transmission drive_type fuel_type  \
5513  6.0L W-12 Gas Turbocharged     Automatic        AWD       Gas   
6761  6.0L W-12 Gas Turbocharged     Automatic        AWD       Gas   
6392               6.2L V-12 Gas     Automatic        AWD       Gas   
1898  6.6L V-12 Gas Turbocharged     Automatic        RWD       Gas   
2647               6.0L V-12 Gas        Manual        RWD       Gas   
434                6.0L V-12 Gas     Automatic        RWD       Gas   
436    3.9L V-8 Gas Turbocharged     Automatic        RWD       Gas   
6255               6.0L V-12 Gas        Manual        RWD       Gas   
1918               6.3L V-12 Gas     Automatic        AWD       Gas   
442    6.8L V-8 Gas Turbocharged     Automatic        RWD       Gas   

     popular_feature    cpo num_accident              usage_type  title  \
5513              []  False            0  Personal or Rental Use  Clean   
6761              []  False            0  Personal or Rental Use  Clean   
6392              []  False            0  Personal or Rental Use  Clean   
1898              []  False            0  Personal or Rental Use  Clean   
2647              []  False            0  Personal or Rental Use  Clean   
434               []  False            0  Personal or Rental Use  Clean   
436               []  False            0  Personal or Rental Use  Clean   
6255              []  False            0  Personal or Rental Use  Clean   
1918              []  False            0  Personal or Rental Use  Clean   
442               []  False            0  Personal or Rental Use  Clean   

     num_owner  
5513         1  
6761         1  
6392         2  
1898         2  
2647         5  
434          5  
436          2  
6255         4  
1918         4  
442          2  

[10 rows x 22 columns]

year       make    model                  sub_model              city  \
779   1999     Suzuki    Grand           4dr JLX Auto 4WD        Manchester   
239   2005      Dodge  Caravan              Grand SXT LWB           Chicago   
4964  2000       Ford   Escort                  ZX2 Coupe           Everett   
4788  1998  Chevrolet    Monte                         LS              Kent   
6733  2001      Honda  Odyssey             EX 7-Passenger  Bloomfield Hills   
4917  2005    Hyundai  Elantra        GLS Sedan Automatic           Orlando   
4381  1998      Dodge  Caravan        Base 3-door FWD SWB              Boaz   
1376  2003      Buick    Regal                         LS          Lakewood   
2611  2002      Honda  Odyssey                         EX      Philadelphia   
351   2004    Hyundai    Santa  GLS 2.7L V6 FWD Automatic   East Landsdowne   

     state  mileage  price            exterior_color interior_color  ...  \
779     MD   180410   1500                     Black           Gray  ...   
239     IL   183245   1500               Stone White        Unknown  ...   
4964    WA   287225   1590     Silver Frost Metallic  Dark Charcoal  ...   
4788    WA   188000   1600                 Torch Red          Black  ...   
6733    MI   233255   1695        Dark Emerald Pearl          Ivory  ...   
4917    FL   191215   1799                       Red        Unknown  ...   
4381    AL   215798   1800                      Blue        Unknown  ...   
1376    NJ   128198   1800  Sterling Silver Metallic    Medium Gray  ...   
2611    PA   215001   1850                      Gray           Gray  ...   
351     PA   195000   1875            Black Obsidian            Tan  ...   

                 engine  transmission drive_type fuel_type  \
779        2.5L V-6 Gas     Automatic        4WD       Gas   
239        3.8L V-6 Gas     Automatic        FWD       Gas   
4964  2.0L Inline-4 Gas        Manual        FWD       Gas   
4788       3.1L V-6 Gas     Automatic        FWD       Gas   
6733       3.5L V-6 Gas     Automatic        FWD       Gas   
4917  2.0L Inline-4 Gas     Automatic        FWD       Gas   
4381       3.0L V-6 Gas     Automatic        FWD       Gas   
1376       3.8L V-6 Gas     Automatic        FWD       Gas   
2611       3.5L V-6 Gas     Automatic        FWD       Gas   
351        2.7L V-6 Gas     Automatic        FWD       Gas   

         popular_feature    cpo num_accident              usage_type  title  \
779                   []  False            0  Personal or Rental Use  Clean   
239   ['Cruise Control']  False            1      Fleet or Mixed Use  Clean   
4964                  []  False            0  Personal or Rental Use  Clean   
4788                  []  False            0  Personal or Rental Use  Clean   
6733  ['Cruise Control']  False            1  Personal or Rental Use  Clean   
4917                  []  False            2  Personal or Rental Use  Clean   
4381                  []  False            0  Personal or Rental Use  Clean   
1376  ['Cruise Control']  False            0  Personal or Rental Use  Clean   
2611  ['Cruise Control']  False            0  Personal or Rental Use  Clean   
351   ['Cruise Control']  False            0  Personal or Rental Use  Clean   

     num_owner  
779          4  
239          4  
4964         3  
4788         6  
6733         1  
4917         1  
4381         3  
1376         4  
2611         3  
351          3  

[10 rows x 22 columns]

In [467]:
df.to_csv('usedCarListingCleaned.csv', encoding = 'utf-8')

After the feature engineering(creating new features and dropping features), dealing with missing values, dropping duplicated rows, there are totally 9702 records and 22 features left.